# Baseline Dictionary

## 1. data load

In [ ]:
!pip install pandas scikit-learn

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd

train_path = '/content/drive/MyDrive/26학년도 1학기/SYNAPSE/난독화된 한글 리뷰 복원 AI 프로젝트/train.csv'
test_path = '/content/drive/MyDrive/26학년도 1학기/SYNAPSE/난독화된 한글 리뷰 복원 AI 프로젝트/test.csv'

train = pd.read_csv(train_path, encoding='utf-8-sig')
test = pd.read_csv(test_path, encoding='utf-8-sig')

print(f"Train 데이터 총 개수: {len(train)}개")
print(f"Test 데이터 총 개수: {len(test)}개")

train.head()

Train 데이터 총 개수: 11263개
Test 데이터 총 개수: 1689개


,ID,input,output
0,TRAIN_00000,별 한 게토 았깝땀. 왜 싸람듯릭 펼 1캐를 쥰눈징 컥꺾폰 싸람믐롯섞 맒록 섧멍핥쟈...,별 한 개도 아깝다. 왜 사람들이 별 1개를 주는지 겪어본 사람으로서 말로 설명하자...
1,TRAIN_00001,잚많 쟉꼬 갉 태 좋눼욥. 차못동 줆 ㅋ,잠만 자고 갈 때 좋네요. 잠옷도 줌 ㅋ
2,TRAIN_00002,절테 간면 않 된는 굣 멥몫,절대 가면 안 되는 곳 메모
3,TRAIN_00003,야... 칵컥 좋꾜 부됴 뼝 뚫렷썹 신원햐쥠만 닮패 넴센 밌쪄벅림. 샥퀘 핥류만 묵...,아... 가격 좋고 뷰도 뻥 뚫려서 시원하지만 담배 냄새 미쳐버림. 싸게 하루만 묵...
4,TRAIN_00004,집윈 축쳐눌료 딴너왓눈뎁 카셩뷔 좋곱 칼쿰한네올. 쩌럼한뒈 뮬콰 욺료토 잊쿄 빻토 ...,지인 추천으로 다녀왔는데 가성비 좋고 깔끔하네요. 저렴한데 물과 음료도 있고 방도 ...


### Train/validation 분리

In [ ]:
from sklearn.model_selection import train_test_split

train_data, val_data = train_test_split(
    train,
    test_size=0.2,
    random_state=42
)

## 2. 단어 사전 만들기

#### train 데이터로만 사전 생성

- zip() 이용해 난독화된 단어(input)과 원문(output)을 쌍으로 묶음
    - *별 한 게토 았깝땀. 왜 싸람듯릭 펼 1캐를 쥰눈징 컥꺾폰...(input) : 별 한 개도 아깝다. 왜 사람들이 별 1개를 주는지 겪어본...(output)*
     
- input과 output을 공백 기준으로 분할 - 토큰화
    - 엷끼카(input) : 여기가(ouput)
- zip()으로 난독화된 단어를 Key, 원문 단어를 Value로 하는 딕셔너리(match_dict)를 생성

In [ ]:
match_dict = {}

# in-out text mapping
for input_text, output_text in zip(
    train_data['input'],
    train_data['output']
):

    # tokenize
    input_words = input_text.split()
    output_words = output_text.split()

    # dictionary
    for iw, ow in zip(input_words, output_words):
        match_dict[iw] = ow

In [ ]:
print(f"{len(match_dict)} , {match_dict}")

141850 , {'엷끼카': '여기가', '억쿠인울룟': '워크인으로', '윕쉴': '입실', '같눙한': '가능한', '곳윈카욤?': '곳인가요?', '없뿔': '어플', '옌악운': '예약은', '않': '안', '됨욤~~': '돼요~~', '췻쑈퇴뉘': '취소되니', '우위샤향울': '유의사항을', '공쥐예': '공지에', '헤놓을셔아': '해놓으셔야', '할': '할', '뜻.': '듯.', '쌩귄': '생긴', '치': '지', '얼먀': '얼마', '앉': '안', '뙨': '된', '홋퉤립랗': '호텔이라', '꺌쿰헹욜': '깔끔해요', '윙칟돛': '위치도', '좋교': '좋고', '묫든': '모든', '쥑건푼듦위': '직원분들이', '친철핥계': '친절하게', '뎃헤': '대해', '쥬셨써': '주셔서', '슉밗한는': '숙박하는', '똥얀': '동안', '깊뷴': '기분', '좋궷': '좋게', '옇헹할': '여행할', '쑤': '수', '위썬썽욘': '있었어요', '따응몌': '다음에', '뿌산': '부산', '어헹': '여행', '옮면': '오면', '또': '또', '엮뀔룟': '여기로', '얘악칼': '예약할', '겆': '것', '갗야올.': '같아요.', '쭝윈창읫': '주인장의', '섶핏수': '서비스', '맛윈두갓': '마인드가', '좋쑵닐따.': '좋습니다.', '찐쩔핫쉭교': '친절하시고', '좋씁뉘타.': '좋습니다.', '맡잇는': '맛있는', '음쉭콰': '음식과', '퍼냔한': '편안한', '꽁간': '공간', '좋앗옷': '좋아요', '~!!': '~!!', '윔찌갸': '위치가', '좋꽃': '좋고', '췬졀햐멱': '친절하며', '깩싫': '객실', '텔피': '대비', '걋썽뷕갓': '가성비가', '좋았숩뉘탑.': '좋았습니다.', '까셩빚': '가성비', '좋운': '좋은', '혹텔뤠써': '호텔에서', '잘': '잘', '쒸콤': '쉬고', '갇오.': '가요.', '햅윤태웨셔': '해운대에서', '콴광할

## 3. 복원 적용



- input으로 들어오는 text 분할, tokenize
- input word와 동일한 단어가 사전에 존재할시 딕셔너리 내의 value로 치환, 미등록 단어일 시 input word 그대로 반환
- join()을 이용해 공백 기준으로 단어 이어 붙임

In [ ]:
def replace_words(input_text, match_dict):
    words = input_text.split()
    replaced_words = [match_dict.get(word, word) for word in words]
    return " ".join(replaced_words)

### validation 복원

In [ ]:
val_base_pred = val_data['input'].apply(
    lambda x: replace_words(x, match_dict)
)

In [ ]:
print(val_base_pred)

6927     온션퓨 가성비 쑥쏘웩오! 별섶 두 펀쳅 빵물닒예옷! 힐링하러 왔는데 완전 힐륑하꾜 ...
1669     밧댜예셔 갖캅귄 한데 수 태뮨엔 전혀 빳타갔 뽀이쥐 않는 굣윕님타. 칵깝임 잊여됴 ...
9499     찜꿀루웨 떡폭뀌윈쮜 낌찧칟캐윈짇 국묾위 묻뗘있곱 찜쿠루 청소는 않옜 안 하니 봐오....
3287     줍번예 룟뒈빼꽈쩜밑 있고, 음식 싸왈섦 편히 잘 놀다 왓엶오. 좋은 횹뗄른 앉닉곬,...
2973     쀼룝 할 일 다 함!! 룸 컨디션 췌샹!! 굴렇냐 옐뤼뻬인떡갔 약간 훈들려 풀않청한...
                               ...                        
5470     출장 올 땐맏댜 묵고 있어요. 얾깆젼키 많이 타넙봤침많 제일 깔끔한 것 카따요. 쭉...
10034    일단 쩐환료 카운터 직원 웅떼 테됴카 퓰퀘헷꼬, 윕찢통 우훙갸 꼴목 않뉘랒 쉭꾸렵교...
8610     한쿱계 챰깐 둘륄 일이 있어 꼬민하댜 껄청한 쑥쇼잎뉘탸. 쩨끄윈욺 하고 나서 방 샹...
6339     읽곧읊 닸십 한번 더 갈 쓱 있다면 열먈를 추떨랏돔 항상 갖섯 줄끽교 쉭콕 냥만을 ...
5333                                 꼬묘툐 호텔 뷰, 윅꾹적귄 방 꽤 좋음
Name: input, Length: 2253, dtype: object


## 4. Accuracy 계산 및 문자 단위 F1 Score 확인

In [ ]:
from collections import Counter

def get_char_f1(y_true, y_pred):
    f1_scores = []

    # 공백 제거 및 문자 단위 토큰화
    for true_text, pred_text in zip(y_true, y_pred):
        true_chars = list(str(true_text).replace(" ", ""))
        pred_chars = list(str(pred_text).replace(" ", ""))

        # 빈 문자열 예외 처리
        if not true_chars and not pred_chars:
            f1_scores.append(1.0)
            continue
        elif not true_chars or not pred_chars:
            f1_scores.append(0.0)
            continue

        # 교집합 문자 개수 계산
        common = Counter(true_chars) & Counter(pred_chars)
        num_same = sum(common.values())

        if num_same == 0:
            f1_scores.append(0.0)
            continue

        # 정밀도(Precision) 및 재현율(Recall) 계산
        precision = num_same / len(pred_chars)
        recall = num_same / len(true_chars)

        # F1 Score 계산
        f1 = 2 * (precision * recall) / (precision + recall)
        f1_scores.append(f1)

    # 평균 F1 스코어 반환
    return sum(f1_scores) / len(f1_scores) if f1_scores else 0.0

In [ ]:
accuracy = (
    val_base_pred == val_data['output']
).mean()

print("Validation Accuracy:", accuracy)

Validation Accuracy: 0.004882379050155349


In [ ]:
charF1 = get_char_f1(val_data['output'], val_base_pred)
print("Validation CharF1:", charF1)

Validation CharF1: 0.4421080374996327


Val Acc는 0.00488 Val Char F1은 0.44210

In [ ]:
from collections import Counter

t_text, p_text = val_data['output'].iloc[0], val_base_pred.iloc[0]
t_chars, p_chars = list(t_text.replace(" ", "")), list(p_text.replace(" ", ""))

# acc 산출
accuracy = 1.0 if t_text == p_text else 0.0

# f1
tp = sum((Counter(t_chars) & Counter(p_chars)).values())
pre = tp / len(p_chars) if p_chars else 0.0
rec = tp / len(t_chars) if t_chars else 0.0
f1 = 2 * (pre * rec) / (pre + rec) if (pre + rec) > 0 else 0.0


# 4. 결과 출력
print(f"정답: {t_text}\n예측: {p_text}")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {pre:.4f} | Recall: {rec:.4f} | F1: {f1:.4f}")

정답: 오션뷰 가성비 숙소에요! 벌써 두 번째 방문이에요! 힐링하러 왔는데 완전 힐링하고 갑니다ㅎㅎ 뷰도 너무 예쁘고 시설도 깔끔하고 직원분들도 친절하세요
예측: 온션퓨 가성비 쑥쏘웩오! 별섶 두 펀쳅 빵물닒예옷! 힐링하러 왔는데 완전 힐륑하꾜 깝뉘댜ㅎㅎ 뷰도 너무 옜푸꼲 시설도 깔끔하고 쥑건뿐둘또 췬졀햐쎄오
Accuracy: 0.0000
Precision: 0.5156 | Recall: 0.5156 | F1: 0.5156


정확도 (Accuracy) 0.0000
- 문장 단위 완전 일치 → acc : 1.0
- 불일치 → acc : 0
- '온션퓨', '쥑건뿐둘또' 등 사전에 없는 미등록 단어를 변환하지 못해 예측 문장과 정답 문장이 완전히 같지 않으므로 0점으로 처리됨.

F1 스코어 (F1 Score) 0.5156
- '가성비', '힐링하러 왔는데 완전', '뷰도 너무', '시설도 깔끔하고' 등 난독화되지 않고 원문과 형태가 동일하게 유지된 문자들이 다수 존재함.
-  전체 글자 중 약 51.5%가 교집합으로 인정되어 부분 점수를 획득함.


**단어 사전(Dictionary) 기반으로 완전환 문장 복원에는 실패하여 acc는 0이지만, 문장에 섞여 있는 정상적인 단어들로 인해 F1스코어는 0.51로 도출**

#### test 데이터 전체를 사전 기반으로 *복원*

In [ ]:
converted_reviews = test['input'].apply(
    lambda x: replace_words(x, match_dict)
).tolist()

## 5. Submission

In [ ]:
submission = pd.read_csv('/content/drive/MyDrive/26학년도 1학기/SYNAPSE/난독화된 한글 리뷰 복원 AI 프로젝트/sample_submission.csv', encoding = 'utf-8-sig')

In [ ]:
submission['output'] = converted_reviews

In [ ]:
submission.to_csv('/content/drive/MyDrive/26학년도 1학기/SYNAPSE/난독화된 한글 리뷰 복원 AI 프로젝트/submission_dictionary.csv', index = False, encoding = 'utf-8-sig')

# Dictionary Baseline 개선

기존 딕셔너리를 이용한 방식
- 난독화된 단어와 딕셔너리에 존재하는 단어가 일치 할때 replace하는 것으로 복원을 진행 했음

**문제**
- 좋아요!, 좋아요~, 좋아요ㅎㅎ 들은 전부 같은 '좋아요'를 가리킴에도 공백 문자 단위로 tokenize 되다보니 다른 단어로 취급
    - 한글, 영문, 숫자, 공백 만 남기고, 단독 자음/모음과 기타 기호들 제거 필요'
    - 원본 사전도 그대로 쓰되, 찾을 수 없을 경우 Cleaned 된 사전ㅇ에서 한번 더 검색

- 쫗아요, 조하요, 좋앝요, 죠하욧 또한 다르게 난독화된 '좋아요'이나 딕셔너리에 '쫗아요' 일부만 등록되어 있을 시 나머지는 '좋아요'로 복원되지 못함
    - 사전에 없는 단어가 들어오더라도 사전 내부에 있는 단어들 중 형태가 비슷한 단어를 찾아 치환 해주는 방식으로 개선 가능

### 1. 노이즈 제거

- https://docs.python.org/ko/3/library/re.html
- 파이썬 정규 표현식 re라이브러리를 이용
- 노이즈 제거 함수 remove_noise()
    -   한글(완성형), 영대소문자, 숫자, 공백을 제외한 모든 문자(특수기호, ㅋㅋ, ㅎㅎ 등) 제거
    

In [ ]:
import re
def remove_noise(word):
    cleaned_word = re.sub(r'[^가-힣a-zA-Z0-9\s]', '', word)
    return cleaned_word

- 기존에는 '좋아요~' , '좋아요...' , ' 좋아요!' 들이 모두 다른 단어로 dictionary에 등록 되어 있으므로 노이즈를 제거 후 dictionary를 재구성 함

In [ ]:
clean_dict = {}

for input_text, output_text in zip(train_data['input'], train_data['output']):
    input_words = input_text.split()
    output_words = output_text.split()

    # 띄어쓰기 불일치로 인한 오류 방지
    if len(input_words) != len(output_words):
        continue

    for iw, ow in zip(input_words, output_words):
        cleaned_iw = remove_noise(iw)

        # 정제 후 빈 문자열이 된 경우(예: 'ㅋㅋ' -> '')는 사전에 넣지 않음
        if cleaned_iw:
            clean_dict[cleaned_iw] = ow

In [ ]:
print(clean_dict)

{'엷끼카': '여기가', '억쿠인울룟': '워크인으로', '윕쉴': '입실', '같눙한': '가능한', '곳윈카욤': '곳인가요?', '없뿔': '어플', '옌악운': '예약은', '않': '안', '됨욤': '돼요~~', '췻쑈퇴뉘': '취소되니', '우위샤향울': '유의사항을', '공쥐예': '공지에', '헤놓을셔아': '해놓으셔야', '할': '할', '뜻': '듯.', '쌩귄': '생긴', '치': '지', '얼먀': '얼마', '앉': '안', '뙨': '된', '홋퉤립랗': '호텔이라', '꺌쿰헹욜': '깔끔해요', '윙칟돛': '위치도', '좋교': '좋고', '묫든': '모든', '쥑건푼듦위': '직원분들이', '친철핥계': '친절하게', '뎃헤': '대해', '쥬셨써': '주셔서', '슉밗한는': '숙박하는', '똥얀': '동안', '깊뷴': '기분', '좋궷': '좋게', '옇헹할': '여행할', '쑤': '수', '위썬썽욘': '있었어요', '따응몌': '다음에', '뿌산': '부산', '어헹': '여행', '옮면': '오면', '또': '또', '엮뀔룟': '여기로', '얘악칼': '예약할', '겆': '것', '갗야올': '같아요.', '쭝윈창읫': '주인장의', '섶핏수': '서비스,', '맛윈두갓': '마인드가', '좋쑵닐따': '좋습니다.', '찐쩔핫쉭교': '친절하시고', '좋씁뉘타': '좋습니다.', '맡잇는': '맛있는', '음쉭콰': '음식과', '퍼냔한': '편안한', '꽁간': '공간', '좋앗옷': '좋아요', '윔찌갸': '위치가', '좋꽃': '좋고', '췬졀햐멱': '친절하며', '깩싫': '객실', '텔피': '대비', '걋썽뷕갓': '가성비가', '좋았숩뉘탑': '좋았습니다.', '까셩빚': '가성비', '좋운': '좋은', '혹텔뤠써': '호텔에서', '잘': '잘', '쒸콤': '쉬고', '갇오': '가요.', '햅윤태웨셔': '해운대에서', '콴광할': '관광할', '땡': '때', '잃곶처굣': '이곳

In [ ]:
def replace_words_perfect(input_text, exact_dict, clean_dict):
    words = input_text.split()
    replaced_words = []

    for word in words:
        # 기호까지 완벽하게 똑같은 단어가 있는지 확인 (문장 부호 보존)
        if word in exact_dict:
            replaced_words.append(exact_dict[word])
        else:
            # 완벽히 똑같지 않다면, 기호를 떼고 핵심 글자만 매칭 시도
            cleaned_word = remove_noise(word)
            if cleaned_word in clean_dict:
                # 기호는 포기하더라도 핵심 글자(오타)라도 복원함
                replaced_words.append(clean_dict[cleaned_word])
            else:
                # 둘 다 없으면 원본 단어 반환
                replaced_words.append(word)

    return " ".join(replaced_words)

In [ ]:
val_rmvN_pred = val_data['input'].apply(
    lambda x: replace_words_perfect(x, match_dict, clean_dict)
)

accuracy = (
    val_rmvN_pred == val_data['output']
).mean()
print("Validation Accuracy:", accuracy)


charF1 = get_char_f1(val_data['output'], val_rmvN_pred)
print("Validation CharF1:", charF1)


Validation Accuracy: 0.004882379050155349
Validation CharF1: 0.4500180711753619


### 2.유사도 기반 매칭 / Fuzzy Matching
쫗아요, 조하요, 좋앝요 등 사전에 없는 단어가 들어왔을 때 포기하지 않고, 사전 내부에 있는 단어들 중 형태가 가장 비슷한 단어를 찾아 Mapping

- 파이썬 내장 라이브러리인 difflib의 유사도 검사(레벤슈타인 거리 기반) 알고리즘을 활용
- 임계치(Cutoff) 이상의 유사도를 가진 단어로 강제 치환


- https://rapidfuzz.github.io/RapidFuzz/ 이용


In [ ]:
!pip install rapidfuzz

In [ ]:
from rapidfuzz import process, fuzz

CLEAN_KEYS = list(clean_dict.keys())

def replace_words_fuzzy(input_text, exact_dict, clean_dict, cutoff=0.7):
    words = input_text.split()
    replaced_words = []

    # 유사도 검사를 위해 사전의 키값 리스트 생성
    clean_keys = list(clean_dict.keys())

    for word in words:
        #  정확히 일치 (Exact Match)
        if word in exact_dict:
            replaced_words.append(exact_dict[word])

        #  노이즈 제거 후 일치 (Cleaned Match)
        else:
            cleaned_word = remove_noise(word)
            if cleaned_word in clean_dict:
                replaced_words.append(clean_dict[cleaned_word])

            # 유사도 기반 매칭 (Fuzzy Match )
            else:
                # 사전에 있는 단어들 중 가장 유사한 단어 찾기
                match = process.extractOne(cleaned_word, CLEAN_KEYS, scorer=fuzz.ratio, score_cutoff=cutoff*100)
                if match:
                    replaced_words.append(clean_dict[match[0]])
                else:
                    replaced_words.append(word) # 복원 실패 시 원본 유지

    return " ".join(replaced_words)

fuzzy 알고리즘을 적용하기 위해 difflib의 get_close_matches()를 이용 했으나 dictionary의 모든 키와 입력 단어(query)를 일일이 파이썬 루프 내에서 비교하므로  시간이 매우 오래걸림.
- 30분 이상 연산이 지속

**c++ 기반 연산의 rapidfuzz 이용**
- https://rapidfuzz.github.io/RapidFuzz/Usage/process.html#extractone
> choices (Collection[Sequence[Hashable]] | Mapping[Sequence[Hashable]]) – list of all strings the query should be compared with or dict with a mapping {<result>: <string to compare>}
- extractOne()를 이용 , process.extractOne 내부의 choices 파라미터에 CLEAN_KEYS를 넘겨줌으로써 빠르게 key 검색
- 약 9분만에 결과 도출

In [ ]:
import os
import json

# 캐시 파일 경로 (json으로 통일)
CACHE_FILE = '/content/drive/MyDrive/26학년도 1학기/SYNAPSE/난독화된 한글 리뷰 복원 AI 프로젝트/fuzzy_results.json'

def get_fuzzy_predictions(val_data, match_dict, clean_dict, cutoff=70):
    # 1. 캐시 파일이 존재하면 로드
    if os.path.exists(CACHE_FILE):
        print("캐시 파일 발견! 저장된 결과를 로드합니다.")
        with open(CACHE_FILE, 'r', encoding='utf-8') as f:
            return json.load(f)

    # 2. 파일이 없을 경우 연산 실행
    print("캐시 파일 없음. 연산을 시작합니다...")
    val_pred_list = [replace_words_fuzzy(x, match_dict, clean_dict, cutoff)
                     for x in val_data['input']]

    # 3. 결과 저장 (JSON 형식)
    with open(CACHE_FILE, 'w', encoding='utf-8') as f:
        json.dump(val_pred_list, f, ensure_ascii=False)

    return val_pred_list

# 실행
val_pred_list = get_fuzzy_predictions(val_data, match_dict, clean_dict)
val_fuzz_pred = pd.Series(val_pred_list, index=val_data.index)

print(val_fuzz_pred)

캐시 파일 발견! 저장된 결과를 로드합니다.
6927     오션 가성비 숙소예요. 별섶 두 펀쳅 빵물닒예옷! 힐링하러 왔는데 완전 힐링하고 갑...
1669     밧댜예셔 갖캅귄 한데 수 태뮨엔 전혀 바다 뽀이쥐 않는 곳입니다. 칵깝임 있어 전혀...
9499     침구류에 떡폭뀌윈쮜 낌찧칟캐윈짇 물이 묻뗘있곱 침구류도 청소는 않옜 안 하니 봐요....
3287     주변 룟뒈빼꽈쩜밑 있고, 음식 싸왈섦 편히 잘 놀다 왓엶오. 좋은 호텔 아니, 그냥...
2973     쀼룝 할 일 다 함. 룸 컨디션 최상. 굴렇냐 옐뤼뻬인떡갔 약간 훈들려 풀않청한 늚...
                               ...                        
5470     출장 올 땐맏댜 묵고 있어요. 얾깆젼키 많이 타넙봤침많 제일 깔끔한 것 가요! 추천...
10034    일단 전환 카운터 직원 웅떼 때도 퓰퀘헷꼬, 윕찢통 우훙갸 꼴목 아니, 시끄럽고, ...
8610     크게 챰깐 둘륄 일이 있어 꼬민하댜 정한 쑥쇼잎뉘탸. 체크인 하고 나서 방 샹땔를 ...
6339     곳을 다시는 한번 더 갈 쓱 있다면 말을 추떨랏돔 항상 갖섯 줄끽교 쉭콕 냥만을 만...
5333                                 꼬묘툐 호텔 뷰, 윅꾹적귄 방 꽤 좋음
Length: 2253, dtype: object


-  매번 동일한 로직으로 연산하므로 결과값 동일
-  매번 연산은 비효율, 한번 계산후 json 파일로 저장.

In [ ]:
accuracy = (
    val_fuzz_pred == val_data['output']
).mean()
print("Validation Accuracy:", accuracy)


charF1 = get_char_f1(val_data['output'], val_fuzz_pred)
print("Validation CharF1:", charF1)


Validation Accuracy: 0.007989347536617843
Validation CharF1: 0.5917188520400655


In [ ]:
print("base===============================")
accuracy = (
    val_base_pred == val_data['output']
).mean()
print("Validation Accuracy:", accuracy)


charF1 = get_char_f1(val_data['output'], val_base_pred)
print("Validation CharF1:", charF1)

print("rmvN===============================")
accuracy = (
    val_rmvN_pred == val_data['output']
).mean()
print("Validation Accuracy:", accuracy)


charF1 = get_char_f1(val_data['output'], val_rmvN_pred)
print("Validation CharF1:", charF1)

print("fuzz===============================")
accuracy = (
    val_fuzz_pred == val_data['output']
).mean()
print("Validation Accuracy:", accuracy)


charF1 = get_char_f1(val_data['output'], val_fuzz_pred)
print("Validation CharF1:", charF1)

base===============================
Validation Accuracy: 0.004882379050155349
Validation CharF1: 0.4421080374996327
rmvN===============================
Validation Accuracy: 0.004882379050155349
Validation CharF1: 0.4500180711753619
fuzz===============================
Validation Accuracy: 0.007989347536617843
Validation CharF1: 0.5917188520400655


# 정리

### Rule-based 방식
**train.csv 데이터를 공백 기준으로 토큰화**
- 난독화 단어(input)를 Key, 원문(output)을 Value로 하는매핑 사전(match_dict)을 구축

**replace_words 함수**
-  통해 입력 문장을 어절 단위로 분할. 사전 내 Key 존재 여부를 확인하여, 존재 시 Value로 치환하고 미존재 시 원본 토큰을 그대로 유지하는 방식을 사용함.

- 입력된 단어와 사전의 Key값이 특수기호, 공백, 글자 구성까지 단 1비트의 오차 없이 100% 동일해야만 복원이 수행되는 가장 엄격한 매칭 방식임.

---

**장점**
- 변환 로직이 명확하여 결과값이 도출되는 과정을 완벽히 추적할 수 있음.
- 복잡한 학습 과정 없이 즉각적인 치환이 가능하여 리소스 소모가 매우 적음.
- 학습 데이터에 존재하는 규칙적인 난독화 패턴에 대해서는 완벽한 복원을 보장함.

**강점을 보이는 난독화 패턴**
- 특정 글자가 항상 특정 자모음으로 치환되는 규칙이 일관된 경우.
- 난독화된 토큰이 원문의 어절 단위와 1:1로 대응되며, 띄어쓰기 위치가 보존된 형태.

---
**한계**


- 띄어쓰기 변화, 오타, 특수문자 조합 등이 조금만 바뀌어도 완전히 다른 토큰으로 인식하여 매칭에 실패함. 유연성이 부족.
    - Dictionary에 정의되지 않은 신조어나 새로운 난독화 패턴이 등장할 경우 대응이 불가능

    - 단어를 독립적으로 처리하므로, 문장 전체의 의미나 주변 단어와의 관계(Context)를 전혀 고려하지 못함.


- 사용자가 생성하는 새로운 단어나 난독화 방식의 조합은 이론적으로 무한대이므로, 이를 모두 사전으로 정의하는 것은 현실적으로 불가능함.
    - 추가 로직이 없는 한 학습 데이터에 없던 새로운 표현이 들어오면 모델은 해당 정보를 "모르는 단어"로 처리할 수밖에 없으며, 이는 일반화 성능의 급격한 하락으로 이어짐.
    - *ex) 인간은 '좋아요'와 '쫗아요'의 의미적 연결을 유추할 수 있으나, 해당 모델은 두 단어를 완전히 별개의 객체로 취급하므로 미등록 패턴을 추론할 능력이 전무함.*
- 퍼지 매칭(Fuzzy matching) 등 유연성을 위한 보완 로직 추가 시, GPU 병렬 처리(Parallel processing) 적용이 어려워 속도 저하 발생.

---
## 성능 요약

| name | 특징 | Validation Accuracy | Validation CharF1 |
| :--- | :---: | :---: | ---: |
| baseline | 기본 dictionary 기반 replace | 0.004882379050155349 | 0.4421080374996327 |
| noise | 특수 기호등의 노이즈 제거 | 0.004882379050155349 | 0.4500180711753619 |
| fuzzy | fuzzy 알고리즘을 이용해 문자열 유사도 검사 | 0.007989347536617843 |  0.5917188520400655|
---

## 전체 동향 분석
**매우 낮은 Accuracy:**
- 3가지 방식 모두 Validation Accuracy가 1% 미만(0.004~0.007)으로, 완벽한 문자열 일치(Exact match)를 달성하는 데 실패하고 있음.

**상대적으로 높은 문자 단위 F1(CharF1)**
- CharF1은 0.44~0.59 수준으로, 완벽히 맞추진 못해도 정답과 상당 부분 유사한 문자 구성을 예측해 내고 있음을 의미.

## 모델별 분석

 **Noise (특수 기호 및 단독 모음/자음 노이즈 제거)**
- Baseline 대비 Accuracy는 완벽히 동일하며, CharF1만 소폭 상승(+0.0079).
- 단순 전처리만으로는 Accuracy 개선이 불가능하며, 이는 완벽하게 문장이 일치하도록 하지는 못함을 의미


**Fuzzy (문자열 유사도 검사)**
- Baseline 대비 Accuracy 약 1.6배 상승(+0.0031), CharF1 대폭 상승(+0.1496).
- 오타, 띄어쓰기, 변형 등을 유연하게 처리하는 Fuzzy 알고리즘 으로  '유연성 부족' 문제를 일부 해소하므로서 좋은 성능 개선을 이끌어냄.